src/porepy/numerics/time_step_control.py

```diff
@@ -0,0 +1,273 @@
+        iter_lowupp_factor: Tuple[float, float] = (1.3, 0.7),
+        recomp_factor: float = 0.5,
+        recomp_max: int = 10,
+        print_info: bool = False,
+    ):
+        """Computes the next time step based on the number of non-linear iterations.
+
+        Parameters:
+            schedule (List): List containing the target times for the simulation.
+                The time-stepping algorithm will adapt the time step so that the target
+                times are guaranteed to be hit/reached. The list must contain minimally
+                two elements, corresponding to the the initial and final simulation times.
+                Lists of length > 2 must contain strictly increasing times.
+                Examples of valid inputs are:
+                  [0, 1], [0, 10, 30, 50], [0, 1*pp.HOUR, 3*pp.HOUR].
+                Examples of invalid inputs are:
+                  [1], [1, 0], [0, 1, 1, 2].
+            dt_init (float): Initial time step.
+            dt_min_max (Tuple of float): Minimum and maximum permissible time steps.
+            iter_max (int): Maximum number of iterations. Default is 15.
+            iter_optimal_range (Tuple of int): Lower and upper optimal iteration range.
+                Default is (4, 7).
+            iter_lowupp_factor (Tuple of float, optional): Lower and upper multiplication
+                factors. Default is (1.3, 0.7).
+            recomp_factor (float). Failed-to-converge recomputation factor.
+                Default is 0.5.
+            recomp_max (int). Failed-to-converge maximum recomputation attempts.
+                Default is 10.
+            print_info (bool). Print time-stepping information. Default is False.
+
+        """
+
+        # Sanity checks on input paramters
+        if len(schedule) < 2:
+            s = "Schedule list must have at least two items, representing the initial and"
+            s += " final simualtion time."
+            raise ValueError(s)
+        elif schedule[0] < 0:
+            raise ValueError("Initial time cannot be negative.")
+        elif schedule[-1] < schedule[0]:
+            raise ValueError("Final time cannot be smaller than initial time.")
+        elif not self._is_strictly_increasing(schedule):
+            raise ValueError("Schedule must contain strictly increasing times.")
+
+        if dt_init <= 0:
+            raise ValueError("Initial time step must be positive")
+        elif dt_init > schedule[-1]:
+            raise ValueError(
+                "Inital time step cannot be larger than final simulation time."
+            )
+        elif dt_init < dt_min_max[0]:
+            raise ValueError(
+                "Intial time step cannot be smaller than minimum time step."
+            )
+        elif dt_init > dt_min_max[1]:
+            raise ValueError(
+                "Intial time step cannot be larger than maximum time step."
+            )
+
+        if dt_min_max[0] > dt_min_max[1]:
```

Comment:
We could also check that the combination of dt_min_max and iter_lowupp_factor are sensible. I.e. that dt_min * iter_low_factor < dt_max and that  dt_max * iter_upp_factor > dt_min

Reply:
I agree. I will add this check. Do you believe it should raise a ValueError or rather a warning?

Reply:
I'm not sure, probably a warning is sufficient.